File: shared/src/commonMain/kotlin/com/aptrade/shared/domain/Portfolio.kt
package com.aptrade.shared.domain

import com.ionspin.kotlin.bignum.decimal.BigDecimal
import com.ionspin.kotlin.bignum.decimal.DecimalMode
import com.ionspin.kotlin.bignum.decimal.RoundingMode

/** Division mode for portfolio math: mirrors Swift Decimal's 38-significant-digit plain
 *  rounding. ionspin BigDecimal THROWS on non-terminating division without a mode. */
val MONEY_MATH = DecimalMode(38, RoundingMode.ROUND_HALF_AWAY_FROM_ZERO)

/** A derived snapshot of a portfolio valued against current quotes. Pure. */
data class PortfolioValuation(
    val cash: Money,
    val holdingsValue: Money,
    val totalValue: Money,
    val unrealizedPnL: Money,
    val dayChange: Money,
)

/** A simulated (paper-trading) portfolio: virtual cash plus average-cost positions and a
 *  transaction log. All transitions are pure and return a new Portfolio. Transcribed from
 *  the Swift original (Sources/APTradeDomain/Portfolio.swift) — semantics must not drift. */
data class Portfolio(
    val cash: Money,
    val positions: List<Position> = emptyList(),
    val transactions: List<Transaction> = emptyList(),
    /** The cash this portfolio OPENED with — fixed at reset time, never moved by a trade.
     *
     *  RECORDED DIVERGENCE FROM SWIFT (M11.2 kickoff decision 4a.1, carry-notes §2.1): the Swift
     *  twin carries this field with NO reader — total return there derives from the equity
     *  curve's own first point, and the reset flow reads `AppSettings.defaultStartingCash`.
     *  Kotlin ports it AND gives it a real consumer: `PerformanceMetrics.sinceInceptionReturn`
     *  (see `FetchPerformanceReport`), which measures return against the balance the user
     *  actually chose rather than against whatever value the priced curve happened to open at.
     *  A $10k practice run and a $1M one must not both report return against the same baseline.
     *  This is a Swift BACKPORT CANDIDATE once the metric proves out here.
     *
     *  Defaults to [cash] so every existing three-argument construction site keeps its meaning:
     *  a portfolio built with no explicit opening balance opened at its current cash. */
    val startingCash: Money = cash,
) {
    fun positionFor(symbol: String): Position? = positions.firstOrNull { it.asset.symbol == symbol }

    /** Epoch-seconds of the EARLIEST transaction — the account's inception instant — or `null`
     *  when nothing has ever traded.
     *
     *  ONE named derivation, deliberately: `FetchPortfolioPerformance`'s `sinceInception` trim
     *  (M11.2 Task 7) and `GoalMath`'s account-age history floor (Task 6) both need exactly this
     *  signal, and the M11 carry-notes require they cannot drift apart. Every consumer calls
     *  here; nobody re-derives `transactions.minOfOrNull { it.epochSeconds }` locally. */
    fun inceptionEpochSeconds(): Long? = transactions.minOfOrNull { it.epochSeconds }

    fun buying(
        asset: Asset,
        quantity: BigDecimal,
        price: Money,
        epochSeconds: Long,
        id: String = generateTradeId(),
        pieId: String? = null,
        isDrip: Boolean = false,
    ): Portfolio {
        if (quantity.isZero()) throw TradeError.InvalidQuantity
        val cost = price.amount * quantity
        if (cash.amount < cost) throw TradeError.InsufficientFunds

        val index = positions.indexOfFirst { it.asset.symbol == asset.symbol }
        val updated = positions.toMutableList()
        if (index >= 0) {
            val old = positions[index]
            val newQty = old.quantity + quantity
            val newAvg = (old.averageCost.amount * old.quantity + cost).divide(newQty, MONEY_MATH)
            updated[index] = Position(old.asset, newQty, Money(newAvg, price.currencyCode), old.realizedPnL)
        } else {
            updated += Position(asset, quantity, price, Money(BigDecimal.ZERO, price.currencyCode))
        }

        val txn = Transaction(id, asset.symbol, TradeSide.Buy, quantity, price, epochSeconds, pieId, isDrip)
        return Portfolio(
            cash = Money(cash.amount - cost, cash.currencyCode),
            positions = updated,
            transactions = transactions + txn,
            startingCash = startingCash,
        )
    }

    /** Credits a dividend payout: cash increases by `shares × amountPerShare` and a
     *  `Dividend` transaction is appended. Positions and cost basis are untouched — this is
     *  a cash event, not a trade. Pure. Transcribed from
     *  `Sources/APTradeDomain/Portfolio.swift:receivingDividend`. */
    fun receivingDividend(
        id: String = generateTradeId(),
        symbol: String,
        amountPerShare: Money,
        shares: BigDecimal,
        exDateEpochSeconds: Long,
    ): Portfolio {
        if (shares <= BigDecimal.ZERO || amountPerShare.amount <= BigDecimal.ZERO) {
            throw TradeError.InvalidQuantity
        }
        val credit = amountPerShare.amount * shares

        val txn = Transaction(id, symbol, TradeSide.Dividend, shares, amountPerShare, exDateEpochSeconds)
        return Portfolio(
            cash = Money(cash.amount + credit, cash.currencyCode),
            positions = positions,
            transactions = transactions + txn,
            startingCash = startingCash,
        )
    }

    fun selling(
        symbol: String,
        quantity: BigDecimal,
        price: Money,
        epochSeconds: Long,
        id: String = generateTradeId(),
        pieId: String? = null,
    ): Portfolio {
        if (quantity.isZero()) throw TradeError.InvalidQuantity
        val index = positions.indexOfFirst { it.asset.symbol == symbol }
        if (index < 0 || positions[index].quantity < quantity) throw TradeError.InsufficientShares

        val old = positions[index]
        val proceeds = price.amount * quantity
        val realizedDelta = (price.amount - old.averageCost.amount) * quantity
        val newQty = old.quantity - quantity

        val updated = positions.toMutableList()
        if (newQty.isZero()) {
            updated.removeAt(index)
        } else {
            updated[index] = Position(
                old.asset, newQty, old.averageCost,
                Money(old.realizedPnL.amount + realizedDelta, old.realizedPnL.currencyCode),
            )
        }

        val txn = Transaction(id, symbol, TradeSide.Sell, quantity, price, epochSeconds, pieId)
        return Portfolio(
            cash = Money(cash.amount + proceeds, cash.currencyCode),
            positions = updated,
            transactions = transactions + txn,
            startingCash = startingCash,
        )
    }

    /** Values every position against `quotes` (cost-basis fallback when a quote is
     *  missing). Day change per share is derived as price − previousClose. Pure. */
    fun valuation(quotes: Map<String, Quote>): PortfolioValuation {
        var holdings = BigDecimal.ZERO
        var unrealized = BigDecimal.ZERO
        var day = BigDecimal.ZERO
        for (position in positions) {
            val q = position.quantity
            val quote = quotes[position.asset.symbol]
            if (quote != null) {
                holdings += quote.price.amount * q
                unrealized += (quote.price.amount - position.averageCost.amount) * q
                day += (quote.price.amount - quote.previousClose.amount) * q
            } else {
                holdings += position.averageCost.amount * q   // cost-basis fallback
            }
        }
        val code = cash.currencyCode
        return PortfolioValuation(
            cash = cash,
            holdingsValue = Money(holdings, code),
            totalValue = Money(cash.amount + holdings, code),
            unrealizedPnL = Money(unrealized, code),
            dayChange = Money(day, code),
        )
    }

    companion object {
        /** The ONE hardcoded opening balance in the shared core (carry-notes §2.7 permits exactly
         *  two literals: this one, and `AppSettings.defaultStartingCash`'s default). Named rather
         *  than inlined so a repo grep for a stray balance finds every real call site instead of
         *  a scatter of bare `100000`s. */
        val DEFAULT_STARTING_CASH: Money = Money.usd("100000")

        /** A fresh paper portfolio opened at [cash], with [startingCash] recorded to match. */
        fun starting(cash: Money = DEFAULT_STARTING_CASH): Portfolio =
            Portfolio(cash = cash, startingCash = cash)
    }
}

File: Sources/APTradeDomain/Portfolio.swift
import Foundation

/// A derived snapshot of a portfolio valued against a set of current quotes. Pure.
public struct PortfolioValuation: Equatable, Sendable {
    public let cash: Money
    public let holdingsValue: Money
    public let totalValue: Money
    public let unrealizedPnL: Money
    public let dayChange: Money

    public init(cash: Money, holdingsValue: Money, totalValue: Money,
                unrealizedPnL: Money, dayChange: Money) {
        self.cash = cash
        self.holdingsValue = holdingsValue
        self.totalValue = totalValue
        self.unrealizedPnL = unrealizedPnL
        self.dayChange = dayChange
    }
}

/// A simulated (paper-trading) portfolio: virtual cash plus average-cost positions and
/// a transaction log. All transitions are pure and return a new `Portfolio`.
public struct Portfolio: Equatable, Codable, Sendable {
    public let cash: Money
    public let positions: [Position]
    public let transactions: [Transaction]

    public init(cash: Money, positions: [Position] = [], transactions: [Transaction] = []) {
        self.cash = cash
        self.positions = positions
        self.transactions = transactions
    }

    /// The starting paper portfolio: $100,000 cash, no holdings.
    public static func starting(cash: Money = Money(amount: 100_000)) -> Portfolio {
        Portfolio(cash: cash)
    }

    public func position(for symbol: String) -> Position? {
        positions.first { $0.asset.symbol == symbol }
    }

    public func buying(_ asset: Asset, quantity: Quantity, at price: Money,
                       on date: Date = Date(), pieId: String? = nil, isDrip: Bool = false) throws -> Portfolio {
        guard !quantity.isZero else { throw TradeError.invalidQuantity }
        let cost = price.amount * quantity.amount
        guard cash.amount >= cost else { throw TradeError.insufficientFunds }

        var updated = positions
        if let index = positions.firstIndex(where: { $0.asset.symbol == asset.symbol }) {
            let old = positions[index]
            let newQty = old.quantity.amount + quantity.amount
            let newAvg = (old.averageCost.amount * old.quantity.amount + cost) / newQty
            updated[index] = Position(
                asset: old.asset,
                quantity: Quantity(newQty),
                averageCost: Money(amount: newAvg, currencyCode: price.currencyCode),
                realizedPnL: old.realizedPnL
            )
        } else {
            updated.append(Position(
                asset: asset,
                quantity: quantity,
                averageCost: price,
                realizedPnL: Money(amount: 0, currencyCode: price.currencyCode)
            ))
        }

        let txn = Transaction(symbol: asset.symbol, side: .buy,
                              quantity: quantity, price: price, date: date, pieId: pieId, isDrip: isDrip)
        return Portfolio(
            cash: Money(amount: cash.amount - cost, currencyCode: cash.currencyCode),
            positions: updated,
            transactions: transactions + [txn]
        )
    }

    /// Credits a dividend payout: cash increases by `shares × amountPerShare` and a
    /// `.dividend` transaction is appended. Positions and cost basis are untouched —
    /// this is a cash event, not a trade. Pure.
    public func receivingDividend(_ symbol: String, amountPerShare: Money,
                                  shares: Quantity, on exDate: Date) throws -> Portfolio {
        guard !shares.isZero, amountPerShare.amount > 0 else { throw TradeError.invalidQuantity }
        let credit = amountPerShare.amount * shares.amount

        let txn = Transaction(symbol: symbol, side: .dividend,
                              quantity: shares, price: amountPerShare, date: exDate)
        return Portfolio(
            cash: Money(amount: cash.amount + credit, currencyCode: cash.currencyCode),
            positions: positions,
            transactions: transactions + [txn]
        )
    }

    public func selling(_ symbol: String, quantity: Quantity, at price: Money,
                        on date: Date = Date(), pieId: String? = nil) throws -> Portfolio {
        guard !quantity.isZero else { throw TradeError.invalidQuantity }
        guard let index = positions.firstIndex(where: { $0.asset.symbol == symbol }),
              positions[index].quantity.amount >= quantity.amount else {
            throw TradeError.insufficientShares
        }

        let old = positions[index]
        let proceeds = price.amount * quantity.amount
        let realizedDelta = (price.amount - old.averageCost.amount) * quantity.amount
        let newQty = old.quantity.amount - quantity.amount

        var updated = positions
        if newQty == 0 {
            updated.remove(at: index)
        } else {
            updated[index] = Position(
                asset: old.asset,
                quantity: Quantity(newQty),
                averageCost: old.averageCost,
                realizedPnL: Money(amount: old.realizedPnL.amount + realizedDelta,
                                   currencyCode: old.realizedPnL.currencyCode)
            )
        }

        let txn = Transaction(symbol: symbol, side: .sell,
                              quantity: quantity, price: price, date: date, pieId: pieId)
        return Portfolio(
            cash: Money(amount: cash.amount + proceeds, currencyCode: cash.currencyCode),
            positions: updated,
            transactions: transactions + [txn]
        )
    }

    /// Values every position against `quotes` (falling back to cost basis when a quote
    /// is missing) and rolls them into account totals. Pure.
    public func valuation(quotes: [String: Quote]) -> PortfolioValuation {
        var holdings = Decimal(0)
        var unrealized = Decimal(0)
        var day = Decimal(0)
        for position in positions {
            let q = position.quantity.amount
            if let quote = quotes[position.asset.symbol] {
                holdings += quote.price.amount * q
                unrealized += (quote.price.amount - position.averageCost.amount) * q
                day += quote.change.amount * q
            } else {
                holdings += position.averageCost.amount * q   // cost-basis fallback
            }
        }
        let code = cash.currencyCode
        return PortfolioValuation(
            cash: cash,
            holdingsValue: Money(amount: holdings, currencyCode: code),
            totalValue: Money(amount: cash.amount + holdings, currencyCode: code),
            unrealizedPnL: Money(amount: unrealized, currencyCode: code),
            dayChange: Money(amount: day, currencyCode: code)
        )
    }
}
